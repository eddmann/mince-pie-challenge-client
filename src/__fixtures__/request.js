export default function createRequestStack(responses = []) {
  const push = ({ path, method, body, response }) => responses.push({ path, method, body, response });

  const api = (path, method, body = undefined) => {
    const actual = { path, method, body };

    if (responses.length === 0) {
      return fail(`Expected: Nil, Actual: ${JSON.stringify(actual)}`);
    }

    const { response, ...expected } = responses.shift();

    expect(expected).toEqual(actual);

    return Promise.resolve(response);
  };

  const upload = (url, file) => api(url, 'PUT', file);

  return { push, api, upload };
};
