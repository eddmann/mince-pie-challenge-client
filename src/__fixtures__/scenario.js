export default function createScenario(fixtures = () => ({})) {
  return (label, run) => {
    const steps = [];

    run({
      given: (label, step) => steps.push(step),
      when: (label, step) => steps.push(step),
      then: (label, step) => steps.push(step),
      fixtures: fixtures(),
    });

    it(label, () =>
      steps.reduce((state, step) => state
        .then(wrapper => step(wrapper) || wrapper)
        .then(wrapper => new Promise(res => process.nextTick(() => { wrapper && wrapper.update(); res(wrapper); }))),
      Promise.resolve(undefined)));
  };
};
