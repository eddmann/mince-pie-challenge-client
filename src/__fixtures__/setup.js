import React from 'react';
import Enzyme, { mount} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import createMemoryHistory from 'history/createBrowserHistory';
import createScenario from './scenario';
import createRequestStack from './request';
import app from '../app';

Enzyme.configure({ adapter: new Adapter() });

export const scenario = createScenario(() => {
  const history = createMemoryHistory();
  const request = createRequestStack();
  const auth = { getToken: () => Promise.resolve() };

  return {
    request,
    auth,
    renderApp: (path = '/') => {
      history.replace(path);
      return app({
        history,
        services: { request, auth },
        links: { add: 'add.url', list: 'list.url', view: 'view.url/{id}' },
      }).then(mount);
    },
  };
});
