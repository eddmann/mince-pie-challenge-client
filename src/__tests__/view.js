import { scenario } from '../__fixtures__/setup';

scenario('view public mince pie', ({ given, when, then, fixtures: { request, renderApp } }) => {
  const ID = '1234';
  const NAME = 'Awesome Mince Pie';

  given('a mince pie resource', () => {
    request.push({
      path: `view.url/${ID}`,
      method: 'GET',
      response: {
        _links: {
          self: { href: `/pies/${ID}` },
        },
        id: ID,
        name: NAME,
        rating: { avg: 0, total: 0 },
        photo: 'photo.url',
        thumbnail: 'thumbnail.url',
        addedAt: new Date('2018-01-01').toISOString(),
      },
    });
  });

  when('visit mince pies view page', wrapper => renderApp(`/pies/${ID}`));

  then('displays the mince pies details', wrapper => {
    expect(wrapper.find('.Pie')).toMatchSnapshot();
  });
});

scenario('view rateable mince pie', ({ given, when, then, fixtures: { request, auth, renderApp } }) => {
  const ID = '1234';
  const NAME = 'Awesome Mince Pie';

  given('an authenticated user with rateable mince pie resource', () => {
    auth.getToken = () => Promise.resolve('SAMPLE_TOKEN');

    request.push({
      path: `view.url/${ID}`,
      method: 'GET',
      response: {
        _links: {
          self: { href: `/pies/${ID}` },
          rate: { href: `/pies/${ID}/rate` },
        },
        id: ID,
        name: NAME,
        rating: { avg: 0, total: 0 },
        photo: 'photo.url',
        thumbnail: 'thumbnail.url',
        addedAt: new Date('2018-01-01').toISOString(),
      },
    });
  });

  when('visit mince pies view page', wrapper => renderApp(`/pies/${ID}`));

  then('displays the mince pies details with rating capabilities', wrapper => {
    expect(wrapper.find('.Pie')).toMatchSnapshot();
    expect(wrapper.find('Rating').length).toEqual(1);
  });
});

scenario('view own mince pie', ({ given, when, then, fixtures: { request, auth, renderApp } }) => {
  const ID = '1234';
  const NAME = 'Awesome Mince Pie';

  given('an authenticated user with own mince pie resource', () => {
    auth.getToken = () => Promise.resolve('SAMPLE_TOKEN');

    request.push({
      path: `view.url/${ID}`,
      method: 'GET',
      response: {
        _links: {
          self: { href: `/pies/${ID}` },
          remove: { href: `/pies/${ID}` },
        },
        id: ID,
        name: NAME,
        rating: { avg: 0, total: 0 },
        photo: 'photo.url',
        thumbnail: 'thumbnail.url',
        addedAt: new Date('2018-01-01').toISOString(),
      },
    });
  });

  when('visit mince pies view page', wrapper => renderApp(`/pies/${ID}`));

  then('displays the mince pies details with removal capabilities', wrapper => {
    expect(wrapper.find('.Pie')).toMatchSnapshot();
    expect(wrapper.find('button.negative').length).toEqual(1);
  });
});

scenario('rate mince pie', ({ given, when, then, fixtures: { request, auth, renderApp } }) => {
  const ID = '1234';
  const NAME = 'Awesome Mince Pie';

  given('an authenticated user with rateable mince pie resource', () => {
    auth.getToken = () => Promise.resolve('SAMPLE_TOKEN');

    request.push({
      path: `view.url/${ID}`,
      method: 'GET',
      response: {
        _links: {
          self: { href: `/pies/${ID}` },
          rate: { href: `/pies/${ID}/rate` },
        },
        id: ID,
        name: NAME,
        rating: { avg: 0, total: 0 },
        photo: 'photo.url',
        thumbnail: 'thumbnail.url',
        addedAt: new Date('2018-01-01').toISOString(),
      },
    });

    request.push({
      path: `/pies/${ID}/rate`,
      method: 'PUT',
      body: { rating: 3 },
      response: {},
    });

    request.push({
      path: `view.url/${ID}`,
      method: 'GET',
      response: {
        _links: {
          self: { href: `/pies/${ID}` },
          rate: { href: `/pies/${ID}/rate` },
        },
        id: ID,
        name: NAME,
        rating: { avg: 3, total: 1 },
        photo: 'photo.url',
        thumbnail: 'thumbnail.url',
        addedAt: new Date('2018-01-01').toISOString(),
      },
    });

    return renderApp(`/pies/${ID}`);
  });

  when('attempt to rate the mince pie', wrapper => {
    wrapper.find('Rating RatingIcon').at(2).simulate('click');
  });

  then('rates the mince pie and updates the details view', wrapper => {
    expect(wrapper.find('CardDescription').text()).toEqual('Rated 3/5 from 1 reviews');
  });
});

scenario('remove mince pie', ({ given, when, then, fixtures: { request, auth, renderApp } }) => {
  const ID = '1234';
  const NAME = 'Awesome Mince Pie';

  given('an authenticated user with own mince pie resource', () => {
    auth.getToken = () => Promise.resolve('SAMPLE_TOKEN');

    request.push({
      path: `view.url/${ID}`,
      method: 'GET',
      response: {
        _links: {
          self: { href: `/pies/${ID}` },
          remove: { href: `/pies/${ID}` },
        },
        id: ID,
        name: NAME,
        rating: { avg: 0, total: 0 },
        photo: 'photo.url',
        thumbnail: 'thumbnail.url',
        addedAt: new Date('2018-01-01').toISOString(),
      },
    });

    request.push({
      path: `/pies/${ID}`,
      method: 'DELETE',
      response: {},
    });

    request.push({
      path: 'list.url',
      method: 'GET',
      response: {
        _links: { self: { href: '/pies' } },
        _embedded: { pies: [] },
        total: 0,
      },
    });

    return renderApp(`/pies/${ID}`);
  });

  when('attempt to remove the mince pie', wrapper => {
    wrapper.find('button.negative').simulate('click');
  });

  then('redirected back to browse listings with mince pie removed', wrapper => {
    expect(wrapper.find('.Browse')).toMatchSnapshot();
    expect(wrapper.find('Card').length).toEqual(0);
  });
});
