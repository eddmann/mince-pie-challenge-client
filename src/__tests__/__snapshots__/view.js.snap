// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`remove mince pie 1`] = `
<div
  className="Browse"
>
  <CardGroup
    doubling={true}
    itemsPerRow={4}
  >
    <div
      className="ui doubling four cards"
    />
  </CardGroup>
</div>
`;

exports[`view own mince pie 1`] = `
Array [
  <Card
    centered={true}
    className="Pie"
    fluid={true}
>
    <div
        className="ui centered fluid card Pie"
        onClick={[Function]}
    >
        <Image
            as="img"
            src="photo.url"
            ui={true}
        >
            <img
                className="ui image"
                src="photo.url"
            />
        </Image>
        <CardContent>
            <div
                className="content"
            >
                <CardHeader
                    content="Awesome Mince Pie"
                >
                    <div
                        className="header"
                    >
                        Awesome Mince Pie
                    </div>
                </CardHeader>
                <CardDescription>
                    <div
                        className="description"
                    >
                        Rated 
                        0
                        /5 from 
                        0
                         reviews
                    </div>
                </CardDescription>
            </div>
        </CardContent>
        <CardContent
            extra={true}
        >
            <div
                className="extra content"
            >
                <Button
                    as="button"
                    floated="right"
                    loading={false}
                    negative={true}
                    onClick={[Function]}
                >
                    <button
                        className="ui negative right floated button"
                        onClick={[Function]}
                        role="button"
                    >
                        Remove
                    </button>
                </Button>
            </div>
        </CardContent>
    </div>
</Card>,
  <div
    className="ui centered fluid card Pie"
    onClick={[Function]}
>
    <Image
        as="img"
        src="photo.url"
        ui={true}
    >
        <img
            className="ui image"
            src="photo.url"
        />
    </Image>
    <CardContent>
        <div
            className="content"
        >
            <CardHeader
                content="Awesome Mince Pie"
            >
                <div
                    className="header"
                >
                    Awesome Mince Pie
                </div>
            </CardHeader>
            <CardDescription>
                <div
                    className="description"
                >
                    Rated 
                    0
                    /5 from 
                    0
                     reviews
                </div>
            </CardDescription>
        </div>
    </CardContent>
    <CardContent
        extra={true}
    >
        <div
            className="extra content"
        >
            <Button
                as="button"
                floated="right"
                loading={false}
                negative={true}
                onClick={[Function]}
            >
                <button
                    className="ui negative right floated button"
                    onClick={[Function]}
                    role="button"
                >
                    Remove
                </button>
            </Button>
        </div>
    </CardContent>
</div>,
]
`;

exports[`view public mince pie 1`] = `
Array [
  <Card
    centered={true}
    className="Pie"
    fluid={true}
>
    <div
        className="ui centered fluid card Pie"
        onClick={[Function]}
    >
        <Image
            as="img"
            src="photo.url"
            ui={true}
        >
            <img
                className="ui image"
                src="photo.url"
            />
        </Image>
        <CardContent>
            <div
                className="content"
            >
                <CardHeader
                    content="Awesome Mince Pie"
                >
                    <div
                        className="header"
                    >
                        Awesome Mince Pie
                    </div>
                </CardHeader>
                <CardDescription>
                    <div
                        className="description"
                    >
                        Rated 
                        0
                        /5 from 
                        0
                         reviews
                    </div>
                </CardDescription>
            </div>
        </CardContent>
    </div>
</Card>,
  <div
    className="ui centered fluid card Pie"
    onClick={[Function]}
>
    <Image
        as="img"
        src="photo.url"
        ui={true}
    >
        <img
            className="ui image"
            src="photo.url"
        />
    </Image>
    <CardContent>
        <div
            className="content"
        >
            <CardHeader
                content="Awesome Mince Pie"
            >
                <div
                    className="header"
                >
                    Awesome Mince Pie
                </div>
            </CardHeader>
            <CardDescription>
                <div
                    className="description"
                >
                    Rated 
                    0
                    /5 from 
                    0
                     reviews
                </div>
            </CardDescription>
        </div>
    </CardContent>
</div>,
]
`;

exports[`view rateable mince pie 1`] = `
Array [
  <Card
    centered={true}
    className="Pie"
    fluid={true}
>
    <div
        className="ui centered fluid card Pie"
        onClick={[Function]}
    >
        <Image
            as="img"
            src="photo.url"
            ui={true}
        >
            <img
                className="ui image"
                src="photo.url"
            />
        </Image>
        <CardContent>
            <div
                className="content"
            >
                <CardHeader
                    content="Awesome Mince Pie"
                >
                    <div
                        className="header"
                    >
                        Awesome Mince Pie
                    </div>
                </CardHeader>
                <CardDescription>
                    <div
                        className="description"
                    >
                        Rated 
                        0
                        /5 from 
                        0
                         reviews
                    </div>
                </CardDescription>
            </div>
        </CardContent>
        <CardContent
            extra={true}
        >
            <div
                className="extra content"
            >
                <Rating
                    clearable="auto"
                    disabled={false}
                    maxRating={5}
                    onRate={[Function]}
                    size="huge"
                >
                    <div
                        className="ui huge rating"
                        onMouseLeave={[Function]}
                        role="radiogroup"
                    >
                        <RatingIcon
                            active={false}
                            aria-checked={false}
                            aria-posinset={1}
                            aria-setsize={5}
                            as="i"
                            index={0}
                            key="0"
                            onClick={[Function]}
                            onMouseEnter={[Function]}
                            selected={false}
                        >
                            <i
                                aria-checked={false}
                                aria-posinset={1}
                                aria-setsize={5}
                                className="icon"
                                onClick={[Function]}
                                onKeyUp={[Function]}
                                onMouseEnter={[Function]}
                                role="radio"
                                tabIndex={0}
                            />
                        </RatingIcon>
                        <RatingIcon
                            active={false}
                            aria-checked={false}
                            aria-posinset={2}
                            aria-setsize={5}
                            as="i"
                            index={1}
                            key="1"
                            onClick={[Function]}
                            onMouseEnter={[Function]}
                            selected={false}
                        >
                            <i
                                aria-checked={false}
                                aria-posinset={2}
                                aria-setsize={5}
                                className="icon"
                                onClick={[Function]}
                                onKeyUp={[Function]}
                                onMouseEnter={[Function]}
                                role="radio"
                                tabIndex={0}
                            />
                        </RatingIcon>
                        <RatingIcon
                            active={false}
                            aria-checked={false}
                            aria-posinset={3}
                            aria-setsize={5}
                            as="i"
                            index={2}
                            key="2"
                            onClick={[Function]}
                            onMouseEnter={[Function]}
                            selected={false}
                        >
                            <i
                                aria-checked={false}
                                aria-posinset={3}
                                aria-setsize={5}
                                className="icon"
                                onClick={[Function]}
                                onKeyUp={[Function]}
                                onMouseEnter={[Function]}
                                role="radio"
                                tabIndex={0}
                            />
                        </RatingIcon>
                        <RatingIcon
                            active={false}
                            aria-checked={false}
                            aria-posinset={4}
                            aria-setsize={5}
                            as="i"
                            index={3}
                            key="3"
                            onClick={[Function]}
                            onMouseEnter={[Function]}
                            selected={false}
                        >
                            <i
                                aria-checked={false}
                                aria-posinset={4}
                                aria-setsize={5}
                                className="icon"
                                onClick={[Function]}
                                onKeyUp={[Function]}
                                onMouseEnter={[Function]}
                                role="radio"
                                tabIndex={0}
                            />
                        </RatingIcon>
                        <RatingIcon
                            active={false}
                            aria-checked={false}
                            aria-posinset={5}
                            aria-setsize={5}
                            as="i"
                            index={4}
                            key="4"
                            onClick={[Function]}
                            onMouseEnter={[Function]}
                            selected={false}
                        >
                            <i
                                aria-checked={false}
                                aria-posinset={5}
                                aria-setsize={5}
                                className="icon"
                                onClick={[Function]}
                                onKeyUp={[Function]}
                                onMouseEnter={[Function]}
                                role="radio"
                                tabIndex={0}
                            />
                        </RatingIcon>
                    </div>
                </Rating>
            </div>
        </CardContent>
    </div>
</Card>,
  <div
    className="ui centered fluid card Pie"
    onClick={[Function]}
>
    <Image
        as="img"
        src="photo.url"
        ui={true}
    >
        <img
            className="ui image"
            src="photo.url"
        />
    </Image>
    <CardContent>
        <div
            className="content"
        >
            <CardHeader
                content="Awesome Mince Pie"
            >
                <div
                    className="header"
                >
                    Awesome Mince Pie
                </div>
            </CardHeader>
            <CardDescription>
                <div
                    className="description"
                >
                    Rated 
                    0
                    /5 from 
                    0
                     reviews
                </div>
            </CardDescription>
        </div>
    </CardContent>
    <CardContent
        extra={true}
    >
        <div
            className="extra content"
        >
            <Rating
                clearable="auto"
                disabled={false}
                maxRating={5}
                onRate={[Function]}
                size="huge"
            >
                <div
                    className="ui huge rating"
                    onMouseLeave={[Function]}
                    role="radiogroup"
                >
                    <RatingIcon
                        active={false}
                        aria-checked={false}
                        aria-posinset={1}
                        aria-setsize={5}
                        as="i"
                        index={0}
                        key="0"
                        onClick={[Function]}
                        onMouseEnter={[Function]}
                        selected={false}
                    >
                        <i
                            aria-checked={false}
                            aria-posinset={1}
                            aria-setsize={5}
                            className="icon"
                            onClick={[Function]}
                            onKeyUp={[Function]}
                            onMouseEnter={[Function]}
                            role="radio"
                            tabIndex={0}
                        />
                    </RatingIcon>
                    <RatingIcon
                        active={false}
                        aria-checked={false}
                        aria-posinset={2}
                        aria-setsize={5}
                        as="i"
                        index={1}
                        key="1"
                        onClick={[Function]}
                        onMouseEnter={[Function]}
                        selected={false}
                    >
                        <i
                            aria-checked={false}
                            aria-posinset={2}
                            aria-setsize={5}
                            className="icon"
                            onClick={[Function]}
                            onKeyUp={[Function]}
                            onMouseEnter={[Function]}
                            role="radio"
                            tabIndex={0}
                        />
                    </RatingIcon>
                    <RatingIcon
                        active={false}
                        aria-checked={false}
                        aria-posinset={3}
                        aria-setsize={5}
                        as="i"
                        index={2}
                        key="2"
                        onClick={[Function]}
                        onMouseEnter={[Function]}
                        selected={false}
                    >
                        <i
                            aria-checked={false}
                            aria-posinset={3}
                            aria-setsize={5}
                            className="icon"
                            onClick={[Function]}
                            onKeyUp={[Function]}
                            onMouseEnter={[Function]}
                            role="radio"
                            tabIndex={0}
                        />
                    </RatingIcon>
                    <RatingIcon
                        active={false}
                        aria-checked={false}
                        aria-posinset={4}
                        aria-setsize={5}
                        as="i"
                        index={3}
                        key="3"
                        onClick={[Function]}
                        onMouseEnter={[Function]}
                        selected={false}
                    >
                        <i
                            aria-checked={false}
                            aria-posinset={4}
                            aria-setsize={5}
                            className="icon"
                            onClick={[Function]}
                            onKeyUp={[Function]}
                            onMouseEnter={[Function]}
                            role="radio"
                            tabIndex={0}
                        />
                    </RatingIcon>
                    <RatingIcon
                        active={false}
                        aria-checked={false}
                        aria-posinset={5}
                        aria-setsize={5}
                        as="i"
                        index={4}
                        key="4"
                        onClick={[Function]}
                        onMouseEnter={[Function]}
                        selected={false}
                    >
                        <i
                            aria-checked={false}
                            aria-posinset={5}
                            aria-setsize={5}
                            className="icon"
                            onClick={[Function]}
                            onKeyUp={[Function]}
                            onMouseEnter={[Function]}
                            role="radio"
                            tabIndex={0}
                        />
                    </RatingIcon>
                </div>
            </Rating>
        </div>
    </CardContent>
</div>,
]
`;
